pkg/resource/container_image.go

```diff
@@ -0,0 +1,426 @@
+			continue
+		}
+		logger.Debug("downloading layer %s", layer.Digest.String())
+
+		layerDir := filepath.Join(path, layer.Digest.Hex())
+		err = os.MkdirAll(layerDir, 0700)
+		if err != nil {
+			return fmt.Errorf("could not create layer directory: %v", err)
+		}
+
+		blobInfo := types.BlobInfo{
+			Digest: layer.Digest,
+			Size:   layer.Size,
+		}
+		layerBlob, _, err := imageSource.GetBlob(ctx, blobInfo, cache)
+		if err != nil {
+			return fmt.Errorf("could not download layer blob: %v", err)
+		}
+
+		err = r.decompress(layerBlob, layer.Digest.Hex(), layer.MediaType, layer.Size)
+		if err != nil {
+			return fmt.Errorf("could not decompress layer: %v", err)
+		}
+		err = layerBlob.Close()
+		if err != nil {
+			return fmt.Errorf("could not close layer: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (r *ContainerImage) writeFile(filename string, content []byte) error {
+	file, err := os.OpenFile(filepath.Join(r.clonePath, filename), os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600) // #nosec G304
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	_, err = file.Write(content)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// The decompression process is a little more involved so separated out.
+func (r *ContainerImage) decompress(t io.Reader, layer string, mediaType string, size int64) error {
+	// The maximum file size should be less than 10x the layer size.
+	size = size * 10
+	path := filepath.Join(r.ClonePath(), layer)
+
+	var tarReader *tar.Reader
+
+	if strings.Contains(strings.ToLower(mediaType), "gzip") {
+		gzReader, err := gzip.NewReader(t)
+		if err != nil {
+			return err
+		}
+		tarReader = tar.NewReader(gzReader)
+		defer gzReader.Close()
+	} else {
+		tarReader = tar.NewReader(t)
+	}
+
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		path, err := sanitizePath(path, header.Name)
+		if err != nil {
+			logger.Error("%v - skipped", err)
+			continue
+		}
+		info := header.FileInfo()
+		if info.IsDir() {
+			if err = os.MkdirAll(path, 0700); err != nil {
+				return err
+			}
+			continue
+		}
+
+		file, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600) // #nosec G304
+		if err != nil {
+			logger.Error("%v", err)
+			continue
+		}
+		defer file.Close()
+
+		// Copy a maximum number of bytes (layer size * 10) so we do not get "bombs". It is unlikely that a file
+		// with significant entropy will be compressed more than 10x. We can review this.
+		n, err := io.CopyN(file, tarReader, size)
+		if err != nil && err != io.EOF {
+			return fmt.Errorf("could not copy file to disk: %v", err)
+		}
+		if n >= size {
+			logger.Warning("copying file %s did not finish due to max file size: %v", file.Name(), err)
+		}
+	}
+	return nil
+}
+
+// ClonePath returns where this repo has been cloned if cloned else ""
+func (r *ContainerImage) ClonePath() string {
+	return r.clonePath
+}
+
+// Depth returns the depth for things that have version control
```

Comment:
We could actually support this if we want. Like I only want to scan the last 3 layers or something. 

Thoughts?

Reply:
We could. Wouldn't be difficult.
Would need to make sure the spec has them in order or consistently.

Reply:
They seem to be a consistent order. I have added some implementation to the repo now.